python/mxnet/gluon/estimator/estimator.py

```diff
@@ -241,16 +240,35 @@ def fit(self, train_data,
             from a data batch and load into contexts(devices)
         """
 
-
         self.epochs = epochs
-        if not batch_size:
-            batch_size = 32 * len(self.context)
+        if isinstance(train_data, gluon.data.DataLoader):
+            num_batches = len(train_data)
+            total_samples = len(train_data._dataset)
+            if total_samples == 0:
+                raise ValueError("DataLoader is Empty. Please refer to gluon.data.DataLoader "
+                                 "for more detail")
+            for dt, _ in train_data:
+                batch_size = dt.shape[0]
+                break
+        else:
+            raise ValueError("Please provide the data as gluon.data.DataLoader")
+
+        if isinstance(self.context, list):
+            if batch_size < len(self.context):
+                raise ValueError("The batch size value is small in comparison to the provided"
+                                 "CPU/GPU context. Please provide the batch size value(power of 2) "
```

Comment:
Should the batch size always be a power of 2 ?

Reply:
Not really. It's more of related to mapping of virtual to physical processor.